truffle/src/main/java/org/jruby/truffle/core/string/StringNodes.java

```diff
@@ -1355,6 +1365,307 @@ private int prevCharHead(Encoding enc, byte[]bytes, int p, int s, int end) {
+                            repl = strCompatAndValidNode.call(frame, string, "str_compat_and_valid", repl, rubyEncoding);
+                            tainted |= isTaintedNode.executeIsTainted(repl);
+                            buf = makeConcatNode.executeMake(buf, rope((DynamicObject)repl), enc);
+                            if (rope((DynamicObject) repl).getCodeRange() == CR_VALID){
+                                cr = CR_VALID;
+                            }
+                        }
+                        p += clen;
+                        p1 = p;
+                        p = StringSupport.searchNonAscii(pBytes, p, e);
+                        if (p == -1) {
+                            p = e;
+                            break;
+                        }
+                    }
+                }
+                if (buf == null) {
+                    if (p == e) {
+//                        setCodeRange(cr);
+                        return nil();
+                    }
+                    buf = RopeConstants.EMPTY_ASCII_8BIT_ROPE;
+                }
+                if (p1 < p) {
+                    buf = makeConcatNode.executeMake(buf, RopeOperations.ropeFromByteList(new ByteList(pBytes, p1, p - p1)), enc);
+                }
+                if (p < e) {
+                    if (repBytes != null) {
+                        buf = makeConcatNode.executeMake(buf, RopeOperations.ropeFromByteList(new ByteList(repBytes, rep, replen)), enc);
+                        if (!rep7bit_p){
+                            cr = CR_VALID;
+                        }
+                    }
+                    else {
+                        repl = yield(frame, (DynamicObject)block, createString(new ByteList(pBytes, p, e - p, enc, true)));//);
+//                        repl = block.yieldSpecific(context, RubyString.newString(runtime, pBytes, p, e - p, enc));
+                        repl = strCompatAndValidNode.call(frame, string, "str_compat_and_valid", repl, rubyEncoding);//EncodingUtils.strCompatAndValid(getContext(), repl, enc);
+                        tainted |= isTaintedNode.executeIsTainted(repl);
+                        buf = makeConcatNode.executeMake(buf, rope((DynamicObject)repl), enc);
+                        if (rope((DynamicObject) repl).getCodeRange() == CR_VALID) {
+                            cr = CR_VALID;
+                        }
+                    }
+                }
+            }
+            else {
+	        /* ASCII incompatible */
+                byte[] pBytes = value.unsafeBytes();
+                int p = value.begin();
+                int e = p + value.getRealSize();
+                int p1 = p;
+                int mbminlen = enc.minLength();
+                if (repl != nil()) {
+                    final Rope replRope = rope((DynamicObject) repl);
+                    final ByteList replValue = RopeOperations.getByteListReadOnly(replRope);
```

Comment:
I don't think you need to convert to a `ByteList` here. You should be able to just call `Rope#getBytesCopy`. The begin value will always be 0 and the byte array will always be the exact length, so you don't need a "real size" method. If you don't need to modify the array at all, then you can just use `Rope#getBytes`.